l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py

```diff
@@ -0,0 +1,21 @@
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import SUPERUSER_ID, api
+
+
+def migrate(cr, registry):
+    with api.Environment.manage():
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        partners = env["res.partner"].search([("category_id", "!=", False)])
+        for partner in partners:
+            category = partner.category_id.filtered(
+                lambda c: 'nace_' in c.get_external_id().get(c.id)
+            )
+            if not category:
+                continue
+            if len(category) > 1:
+                category = category[0]
+            partner.nace_id = env.ref(
+                category.get_external_id().get(category.id).replace('old_', '')
+            )
```

Comment:
This is possibly a bottleneck if you have lots of partners. Try this:

```suggestion
# Copyright 2019 ACSONE SA/NV
# Copyright 2020 Tecnativa - Jairo Llopis
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).

from openupgradelib import openupgrade

from odoo import SUPERUSER_ID, api


@openupgrade.migrate()
def migrate(env):
    nace_categories = env['ir.model_data'].search([
        ('module', '=', 'l10_eu_nace'),
        ('model', '=', 'res.partner.category'),
    ])
    for category_imd in nace_categories:
        partners = env['res.partner'].search([
            ('category_id', 'in', [category_imd.res_id]),
        )
        partners.write({
            'nace_id': env.ref('l10n_eu_nace.%s' % category_imd.name.replace('old_', '', 1)).id,
        })
```

Benefits:

1. Single search for all categories and their XMLID names.
1. Write all partners in a category in batch.
1. More readable.